R/integrate_profile.R

```diff
@@ -269,44 +289,56 @@ integrate_profile.vpts <- function(x, alt_min = 0, alt_max = Inf,
 
-  index <- which(x$height >= alt_min & x$height < alt_max)
+  # dh is a vector of the height of each bin used for the computation in km
+  # Its value is zeros for bins completly below alt_min or above alt_max, the
+  # interval value for all bin completly above alt_min and below alt_max, and
+  # a ratio of interval for the bins overlaping alt_min and alt_max
+  dh <- pmin(pmin(pmax(x$height-alt_min,0),interval),
+             pmin(pmax(alt_max-x$height,0),interval)) / 1000
+
+  # Vertically Integrated Density in individuals/km^2
+  vid <- colSums(get_quantity(x, "dens") * dh, na.rm = TRUE)
+  # Vertically Integrated Reflectivity in cm^2/km^2
+  vir <- colSums(get_quantity(x, "eta") * dh, na.rm = TRUE)
+
+  # Projection of migratory direction (alpha) on the flight direction (dd)
   if (is.na(alpha)) {
-    cosfactor <- 1 + 0 * get_quantity(x, "dd")[index, ]
+    cosfactor <- 1 + 0 * get_quantity(x, "dd")
   } else {
-    cosfactor <- cos((get_quantity(x, "dd")[index, ] - alpha) * pi / 180)
+    cosfactor <- cos((get_quantity(x, "dd") - alpha) * pi / 180)
   }
 
-  dens_quantity <- colSums(get_quantity(x, "dens")[index, ], na.rm = TRUE)
+  # Migration Traffic Rate in individuals/km/h
+  # multiply speeds (ff) by 3.6 to convert m/s to km/h
+  mtr <- colSums(get_quantity(x, "dens") * cosfactor * get_quantity(x, "ff")
+             * 3.6  * dh, na.rm = TRUE)
+  # Reflectivity Traffic Rate in cm^2/km/h
+  rtr <- colSums(get_quantity(x, "eta") * cosfactor * get_quantity(x, "ff")
+             * 3.6 * dh, na.rm = TRUE)
+
+  # The following quantites are vertically summed by weighting each altitudinal
+  # bin based on their bird reflectivity value (eta) and the height
+  w <- get_quantity(x, "eta") * dh
+  w[is.na(w)] <- 0
+  w <- sweep(w, 2, colSums(w), FUN="/")
 
-  # multiply speeds by 3.6 to convert m/s to km/h
-  mtr <- colSums(cosfactor * get_quantity(x, "ff")[index, ] * 3.6 *
-    get_quantity(x, "dens")[index, ], na.rm = TRUE) * interval / 1000
-  rtr <- colSums(cosfactor * get_quantity(x, "ff")[index, ] * 3.6 *
-    get_quantity(x, "eta")[index, ], na.rm = TRUE) * interval / 1000
-  vid <- dens_quantity * interval / 1000
-  vir <- colSums(get_quantity(x, "eta")[index, ], na.rm = TRUE) * interval / 1000
-  height <- colSums((x$height[index] + x$attributes$where$interval / 2) *
-    get_quantity(x, "dens")[index, ],
-  na.rm = TRUE
-  ) / dens_quantity
-  u <- colSums(get_quantity(x, "u")[index, ] * get_quantity(x, "dens")[index, ],
-    na.rm = TRUE
-  ) / dens_quantity
-  v <- colSums(get_quantity(x, "v")[index, ] * get_quantity(x, "dens")[index, ],
-    na.rm = TRUE
-  ) / dens_quantity
-  ff <- colSums(get_quantity(x, "ff")[index, ] * get_quantity(x, "dens")[index, ],
-    na.rm = TRUE
-  ) / dens_quantity
+  height <- colSums( (get_quantity(x, "height") + interval / 2) * w, na.rm = T)
```

Comment:
More a general comment: please verify that whenever you place the radar at the bottom of a bin, you retrieve identical results with this modified code

Reply:
The tests not completing is because the changes to the function introduces differences in the plots. If it turns out to be small numerical differences due to the lower bin then we can update the snaps for those images to the new format. You need R 4.1 and vdiffr package installed to review the changes introduced in the plot, on older R versions the image tests are skipped, which is probably why you didn't have problems with `devtools::check()` locally

Reply:
> I don't understand the interval/2 statements here, doesn't that become obsolete once you include integration with sub-bin resolution?
As `get_quantity(x, "height")` is the lowest edge of the elevation bin (0, 200 etc...), adding `interval/2` allows to compute the mean bin elevation. This would already implemented before (i.e., I didn't change anything there).

> I don't see why it should be present in u,v and ff
very true, I think this is a copy/paste error. Corrected in 605db6af951f179ecfe83c922d1c46a9ab443100

Reply:
> More a general comment: please verify that whenever you place the radar at the bottom of a bin, you retrieve identical results with this modified code

Yes, I just checked that and the result are the same for the vp and vpts example.